r1cs/gadgets/std/src/fields/nonnative/reduce.rs

```diff
@@ -223,6 +223,62 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
         Ok(())
     }
 
+    pub(crate) fn pre_mul_reduce_for_pseudomersenne<CS: ConstraintSystemAbstract<ConstraintF>>(
+        mut cs: CS,
+        elem: &mut NonNativeFieldGadget<SimulationF, ConstraintF>,
+        elem_other: &mut NonNativeFieldGadget<SimulationF, ConstraintF>,
+    ) -> Result<(), SynthesisError> {
+        debug_assert!(
+            elem.check() && elem_other.check(),
+            "pre_mul_reduce(): check() failed on input gadgets"
+        );
+
+        let params = get_params(SimulationF::size_in_bits(), ConstraintF::size_in_bits(), SimulationF::Params::DIFFERENCE_WITH_HIGHER_POWER_OF_TWO);
+        // To assure that the limbs of the product representation do not exceed the capacity
+        // bound, we demand
+        // ``
+        //      bits_per_limb + surfeit(product) <= CAPACITY,
+        // ``
+        // where
+        // ``
+        //      surfeit(product) = log(num_limbs * (num_add(L)+1) * (num_add(R) + 1) * (c+1) * 2^bits_per_limb).
+        // ``
```

Comment:
It should be * (c*2^h+1) if we allow that bits_per_limbs does not divide MODULUS_BITS. Correct also below.